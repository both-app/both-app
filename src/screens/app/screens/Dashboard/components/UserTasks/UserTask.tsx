import React, { FC, useContext } from 'react'
import { StyleSheet, Alert } from 'react-native'

import { CardButton } from 'library/components/CardButton'

import { colors } from 'res/colors'

import { TaskContext } from 'screens/app/contexts/Task.context'
import { UsersContext } from 'screens/app/contexts/Users.context'
import { CategoryContext } from 'screens/app/contexts/Category.context'
import { UserTaskContext } from 'screens/app/contexts/UserTask.context'

interface UserTaskProps {
  userTask: UserTask
}

export const UserTask: FC<UserTaskProps> = ({ userTask }) => {
  const { getTaskById } = useContext(TaskContext)
  const { getUserById } = useContext(UsersContext)
  const { getCategoryById } = useContext(CategoryContext)
  const { deleteUserTask } = useContext(UserTaskContext)

  const task = getTaskById(userTask?.taskId)
  const category = getCategoryById(task?.categoryId)
  const user = getUserById(userTask.userId)

  if (!userTask || !task || !category || !user) {
    return null
  }

  const handleOnLongPress = () => {
    Alert.alert('Vous voulez supprimer la tÃ¢che?', '', [
      {
        text: 'Non',
        style: 'cancel',
      },
      {
        text: 'Oui',
        style: 'destructive',
        onPress: () => deleteUserTask(userTask.id),
      },
    ])
  }

  return (
    <CardButton
      emoji={task.emoji}
      title={task.name}
      subtitle={`Par ${user.firstName}`}
      onLongPress={handleOnLongPress}
      points={task.points}
      containerStyle={{
        backgroundColor: category.color,
      }}
      textStyle={styles.cardText}
    />
  )
}

const styles = StyleSheet.create({
  cardText: {
    color: colors.white,
  },
})
