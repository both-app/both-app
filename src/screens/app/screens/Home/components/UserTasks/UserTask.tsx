import React, { FC, useContext } from 'react'
import { StyleSheet, Animated, Text } from 'react-native'
import * as Haptics from 'expo-haptics'

import { CardButton } from 'library/components/CardButton'

import { colors, lightenDarkenColor } from 'res/colors'
import { useT } from 'res/i18n'

import { Point } from 'library/components/Point'

import { TaskContext } from 'screens/app/contexts/Task.context'
import { UsersContext } from 'screens/app/contexts/Users.context'
import { CategoryContext } from 'screens/app/contexts/Category.context'
import { UserTaskContext } from 'screens/app/contexts/UserTask.context'
import { UserScoreContext } from 'screens/app/contexts/UserScore.context'
import { TouchableOpacity } from 'react-native-gesture-handler'
import Swipeable from 'react-native-gesture-handler/Swipeable'
import { IconName, Icon } from 'library/components/Icon'

interface UserTaskProps {
  userTask: UserTask
}

export const UserTask: FC<UserTaskProps> = ({ userTask }) => {
  const { t } = useT()
  const { me } = useContext(UsersContext)
  const { getTaskById } = useContext(TaskContext)
  const { getUserById } = useContext(UsersContext)
  const { getCategoryById } = useContext(CategoryContext)
  const { deleteUserTask } = useContext(UserTaskContext)
  const { fetchUserScore } = useContext(UserScoreContext)

  const task = getTaskById(userTask.taskId)
  const category = getCategoryById(task.categoryId)
  const user = getUserById(userTask.userId)

  if (!user || !category || !task) {
    return null
  }

  const isDeletable =
    userTask.taskId !== 'create_both' &&
    userTask.taskId !== 'join_both' &&
    userTask.userId === me.id

  const onDeletePress = async () => {
    await Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light)
    await deleteUserTask(userTask.id)
    await fetchUserScore()
  }

  const renderRightActions = (progress) => {
    const trans = progress.interpolate({
      inputRange: [0, 1],
      outputRange: [140, 0],
    })
    return (
      <Animated.View style={{ transform: [{ translateX: trans }] }}>
        <TouchableOpacity style={styles.deleteCard} onPress={onDeletePress}>
          <Icon
            iconName={'trash' as IconName}
            width={24}
            height={24}
            style={styles.trashIcon}
          />
          <Text style={styles.deleteText}>{t('app:screen:home:delete')}</Text>
        </TouchableOpacity>
      </Animated.View>
    )
  }

  return (
    <Swipeable renderRightActions={isDeletable && renderRightActions}>
      <CardButton
        emoji={task.emoji}
        title={task.name}
        subtitle={t('app:screen:home:userTask:subtitle', {
          firstName: user.firstName,
        })}
        activeBackgroundColor={lightenDarkenColor(category.color, 10)}
        activeTextColor={colors.white}
        containerStyle={{
          backgroundColor: category.color,
          marginTop: 8,
          marginHorizontal: 24,
        }}
        textStyle={styles.cardText}
        rightContent={<Point points={userTask.points} />}
        disabled={!isDeletable}
      />
    </Swipeable>
  )
}

const styles = StyleSheet.create({
  cardText: {
    color: colors.white,
  },
  deleteCard: {
    minHeight: 64,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginRight: 24,
    marginTop: 8,
    paddingVertical: 20,
    paddingHorizontal: 16,
    borderRadius: 8,
    backgroundColor: colors.skin200,
  },
  trashIcon: {
    color: colors.crtical,
  },
  deleteText: {
    color: colors.crtical,
    fontWeight: '500',
    fontSize: 14,
    marginLeft: 8,
  },
})
