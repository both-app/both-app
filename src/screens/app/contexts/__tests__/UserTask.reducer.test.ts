import { userTaskReducer, State } from '../UserTask.reducer'

const initialState: State = {
  allIds: ['A'],
  byId: {
    A: {
      id: 'A',
      userId: 'userIdA',
      taskId: 'taskIdA',
      createdAt: '2020-04-06T18:52:08.085Z',
      relationId: 'relationIdA',
    },
  },
}

describe('UserTaskReducer', () => {
  test('pushAllUserTasks', () => {
    const result = userTaskReducer(initialState, {
      type: 'pushAllUserTasks',
      userTasks: [
        {
          id: 'B',
          userId: 'userIdA',
          createdAt: '2020-04-06',
          taskId: 'taskId1',
          relationId: 'relationIdA',
        },
        {
          id: 'C',
          userId: 'userIdB',
          createdAt: '2020-04-06',
          taskId: 'taskId1',
          relationId: 'relationIdA',
        },
      ],
    })

    expect(result).toMatchSnapshot()
  })

  test('pushUserTask', () => {
    const result = userTaskReducer(initialState, {
      type: 'pushUserTask',
      userTask: {
        id: 'B',
        userId: 'userIdA',
        createdAt: '2020-04-06',
        taskId: 'taskId1',
        relationId: 'relationIdA',
      },
    })

    expect(result).toMatchSnapshot()
  })

  test('deleteUserTask', () => {
    const result = userTaskReducer(initialState, {
      type: 'deleteUserTask',
      userTaskId: 'A',
    })

    expect(result).toMatchSnapshot()
  })
})
