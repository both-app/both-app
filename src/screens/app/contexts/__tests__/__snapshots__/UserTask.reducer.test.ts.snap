// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTaskReducer deleteUserTask 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {
    "A": Object {
      "createdAt": "2020-04-06T18:52:08.085Z",
      "id": "A",
      "relationId": "relationIdA",
      "taskId": "taskIdA",
      "userId": "userIdA",
    },
  },
}
`;

exports[`UserTaskReducer pushAllUserTasks 1`] = `
Object {
  "allIds": Array [
    "B",
    "C",
  ],
  "byId": Object {
    "B": Object {
      "createdAt": "2020-04-06",
      "id": "B",
      "relationId": "relationIdA",
      "taskId": "taskId1",
      "userId": "userIdA",
    },
    "C": Object {
      "createdAt": "2020-04-06",
      "id": "C",
      "relationId": "relationIdA",
      "taskId": "taskId1",
      "userId": "userIdB",
    },
  },
}
`;

exports[`UserTaskReducer pushUserTask 1`] = `
Object {
  "allIds": Array [
    "B",
    "A",
  ],
  "byId": Object {
    "A": Object {
      "createdAt": "2020-04-06T18:52:08.085Z",
      "id": "A",
      "relationId": "relationIdA",
      "taskId": "taskIdA",
      "userId": "userIdA",
    },
    "B": Object {
      "createdAt": "2020-04-06",
      "id": "B",
      "relationId": "relationIdA",
      "taskId": "taskId1",
      "userId": "userIdA",
    },
  },
}
`;
